test/app-luatest/protobuf_repeated_test.lua

```diff
@@ -0,0 +1,108 @@
+local t = require('luatest')
+local protobuf = require('protobuf')
+local g = t.group()
+
+local p = t.group('packed_repeated_encoding', {
+    {type = 'repeated float', value = {0.5, 1.5},
+        res = '0a080000003f0000c03f'},
+    {type = 'repeated double', value = {0.5, 1.5},
+        res = '0a10000000000000e03f000000000000f83f'},
+    {type = 'repeated int32', value = {1, 2, 3, 4}, res = '0a0401020304'},
+    {type = 'repeated sint32', value = {1, 2, 3, 4}, res = '0a0402040608'},
+    {type = 'repeated uint32', value = {1, 2, 3, 4}, res = '0a0401020304'},
+    {type = 'repeated int64', value = {1, 2, 3, 4}, res = '0a0401020304'},
+    {type = 'repeated sint64', value = {1, 2, 3, 4}, res = '0a0402040608'},
+    {type = 'repeated uint64', value = {1, 2, 3, 4}, res = '0a0401020304'},
+    {type = 'repeated fixed32', value = {1, 2}, res = '0a080100000002000000'},
+    {type = 'repeated sfixed32', value = {1, 2}, res = '0a080100000002000000'},
+    {type = 'repeated fixed64', value = {1, 2},
+        res = '0a1001000000000000000200000000000000'},
+    {type = 'repeated sfixed64', value = {1, 2},
+        res = '0a1001000000000000000200000000000000'},
+    {type = 'repeated bool', value = {true, true}, res = '0a020101'},
+})
+
+p.test_packed_repeated_encoding = function(cg)
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {cg.params.type, 1}})
+    })
+    local result = protocol:encode('test', {val = cg.params.value})
+    t.assert_equals(string.hex(result), cg.params.res)
+end
+
+g.test_packed_repeated_int32_long_tag = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'repeated int32', 1000}})
+    })
+    local result = protocol:encode('test', {val = {1, 2, 3, 4}})
+    t.assert_equals(string.hex(result), 'c23e0401020304')
+end
+
+g.test_packed_repeated_int32_with_default_value = function()
+    local protocol = protobuf.protocol({
+        protobuf.message('test', {val = {'repeated int32', 1}})
+    })
+    local result = protocol:encode('test', {val = {1, 0, 0, 4}})
+    t.assert_equals(string.hex(result), '0a020104')
+end
```

Comment:
This is an interesting test case.

We can't correctly encode `{1, nil, nil, 4}`, so an error about a holey array is raised.

We can't correctly encode `{1, box.NULL, box.NULL, 4}`, so an error should be raised in this case too.

In fact, we can't correctly encode `{1, 0, 0, 4}` (it is not `{1, 4}`). It seems, we should raise an error in this case too?

Technically, we can fallback to the non-packed encoding in these cases. Is it a correct behavior for an encoder? Or an error is expected?

Reply:
Why do we have to raise error in `{1, 0, 0, 4}` case? 0 is a normal value (it's default but it is a value) so encoding is correct?

Reply:
A decoder interprets the result as `{1, 4}` and it is not the same as the source value: `{1, 0, 0, 4}`. If we are obligated to use the packed encoding for the array and the 'no presence' discipline for its items, the array with default values can't be encoded in a way that allows to get it back after decoding. I consider it as inability to correctly encode the value.

Technically we can:

1. Use the non-packed encoding.
2. Use the 'explicit presence' discipline.
3. Give an error.
4. Encode an array anyway.

I don't know what other encoders do and propose to investigate this topic.

Reply:
I didn`t find any information about this type of encoding but I agree that `{1, 4}` is not the same `{1, 0, 0, 4}`. I made an error by this case and fixed tests where it emerged.